File: Tema 1/Error3.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

int main() {
    // 1. El Problema: Error de redondeo en punto flotante
    double a = 0.1 + 0.1 + 0.1;
    double b = 0.3;

    std::cout << std::setprecision(17); // Mostrar toda la precisión
    std::cout << "Valor de a (0.1 * 3): " << a << std::endl;
    std::cout << "Valor de b:           " << b << std::endl;

    // 2. Comparación directa (FALLA)
    std::cout << "\n--- Comparación con '==' ---" << std::endl;
    if (a == b) {
        std::cout << "Resultado: Son iguales" << std::endl;
    } else {
        std::cout << "Resultado: SON DIFERENTES (Error esperado)" << std::endl;
    }

    // 3. La Solución: Uso de Épsilon
    double epsilon = 0.00001;

    std::cout << "\n--- Comparación con Épsilon (" << epsilon << ") ---" << std::endl;
    if (std::abs(a - b) < epsilon) {
        std::cout << "Resultado: Son iguales (dentro del margen de error)" << std::endl;
    } else {
        std::cout << "Resultado: Son diferentes" << std::endl;
    }

    return 0;
}

File: Tema 1/Error3.java
public class Error3 {
    public static void main(String[] args) {
        // 1. El Problema: Error de redondeo en aritmética de punto flotante
        double a = 0.1 + 0.1 + 0.1;
        double b = 0.3;

        System.out.println("Valor de a (0.1 * 3): " + a);
        System.out.println("Valor de b:           " + b);

        // 2. Intento de comparación directa (FALLARÁ)
        System.out.println("\n--- Comparación con '==' ---");
        if (a == b) {
            System.out.println("Resultado: Son iguales");
        } else {
            System.out.println("Resultado: SON DIFERENTES (Error esperado)");
        }

        // 3. La Solución: Uso de un margen de error (Épsilon)
        double epsilon = 0.00001; // Definimos la tolerancia
        
        System.out.println("\n--- Comparación con Épsilon (" + epsilon + ") ---");
        if (Math.abs(a - b) < epsilon) {
            System.out.println("Resultado: Son iguales (dentro del margen de error)");
        } else {
            System.out.println("Resultado: Son diferentes");
        }
    }
}